packages/widget-playground/src/components/DrawerControls/CodeControl/CodeEditor.tsx

```diff
@@ -0,0 +1,176 @@
+import { useEffect, useRef, useState } from 'react';
+import { CircularProgress, Tooltip, useTheme } from '@mui/material';
+import ContentCopyIcon from '@mui/icons-material/ContentCopy';
+import type { BeforeMount, OnMount, OnChange } from '@monaco-editor/react';
+import Editor from '@monaco-editor/react';
+import type { WidgetConfig } from '@lifi/widget';
+import { useThemeMode } from '../../../hooks';
+import { tooltipPopperZIndex } from '../DrawerControls.style';
+import {
+  CodeContainer,
+  EditorContainer,
+  CodeCopyButton,
+} from './CodeControl.style';
+import { getWhitelistedConfig, useConfig } from '../../../store';
+import { getValueFromPath } from '../../../utils';
+
+interface MonacoEditor {
+  layout: (dimensions: { width: number; height: number }) => void;
+}
+
+const configTemplate = (config?: string) =>
+  config ? `const config = ${config}` : null;
+
+const substitions = {
+  walletConfig: {
+    '"walletConfig": {}': '"walletConfig": { async onConnect() {} }',
+  },
+};
+const configToStringWithSubstitions = (
+  config?: Partial<WidgetConfig>,
+): string | undefined => {
+  if (!config) {
+    return undefined;
+  }
+  let stringifiedConfig = JSON.stringify(config, null, 2);
+
+  Object.entries(substitions).forEach(([property, substition]) => {
+    if (getValueFromPath(config, property)) {
+      const [[find, replace]] = Object.entries(substition);
+      stringifiedConfig = stringifiedConfig.replace(find, replace);
+    }
+  });
+
+  return stringifiedConfig.replace(/"([^"]+)":/g, '$1:');
+};
+
+interface CodeEditorProps {
+  onChange?: (code: string | undefined) => void;
+}
+
+export const CodeEditor = ({ onChange }: CodeEditorProps) => {
+  const { config } = useConfig();
+
+  const [editorContent, setEditorContent] = useState('');
+  const [editor, setEditor] = useState();
+  const editorContainerRef = useRef(null);
+  const theme = useTheme();
+  const themeMode = useThemeMode();
+
+  const code = config
+    ? configTemplate(
+        configToStringWithSubstitions(getWhitelistedConfig(config)),
+      )
+    : null;
+
+  useEffect(() => {
+    if (code) {
+      setEditorContent(code);
+    }
+  }, [code]);
```

Comment:
This `useEffect` part looks unnecessary and maybe we can use `code` just fine and avoid re-render 🙂 Wdyt?

Reply:
Have remove this and passed the code directly through to the CodeEditor